tests/Aspire.Components.Common.Tests/ActivityNotifier.cs

```diff
@@ -11,19 +13,44 @@ namespace Aspire.Components.Common.Tests;
 /// </summary>
 public sealed class ActivityNotifier : BaseProcessor<Activity>
 {
-    // RunContinuationsAsynchronously because OnEnd gets invoked on the thread creating the Activity.
-    // Running more test code on this thread can cause deadlocks in the case where the Activity is created on a "drain thread" (ex. Redis)
-    // and the test method disposes the Host on that same thread. Disposing the Host will dispose the Instrumentation, which tries joining
-    // with the "drain thread".
-    private readonly TaskCompletionSource _taskSource = new TaskCompletionSource(TaskCreationOptions.RunContinuationsAsynchronously);
+    private readonly Channel<Activity> _activityChannel = Channel.CreateUnbounded<Activity>();
 
-    public Task ActivityReceived => _taskSource.Task;
+    public async Task<List<Activity>> TakeAsync(int count, TimeSpan timeout)
+    {
+        var activityList = new List<Activity>();
+        using var cts = new CancellationTokenSource(timeout);
+        await foreach (var activity in WaitAsync(cts.Token))
+        {
+            activityList.Add(activity);
+            if (activityList.Count == count)
+            {
+                break;
+            }
+        }
 
-    public List<Activity> ExportedActivities { get; } = [];
+        return activityList;
+    }
 
     public override void OnEnd(Activity data)
     {
-        ExportedActivities.Add(data);
-        _taskSource.SetResult();
+        _activityChannel.Writer.TryWrite(data);
+    }
+
+    private async IAsyncEnumerable<Activity> WaitAsync([EnumeratorCancellation] CancellationToken cancellationToken)
```

Comment:
Why not have `_activityChannel.Reader.ReadAllAsync(cancellationToken).ConfigureAwait(false)` directly in TakeAsync?

Reply:
At first I only had `WaitAsync` public on this class. But when I used it in 4 different places, all consumers needed was to get a collection of the activities and assert on them. So I made the helper `TakeAsync` method and made the `WaitAsync` method private.